File: src/main/java/org/example/tree/BalancedBinaryTree.java
package org.example.tree;

public class BalancedBinaryTree {

    /*public boolean isBalanced(TreeNode root) {
        return dfs(root).getKey();
    }

    private Pair<Boolean, Integer> dfs(TreeNode node) {
        if (node == null) return new Pair<Boolean, Integer>(true, 0);

        Pair<Boolean, Integer> left = dfs(node.left);
        Pair<Boolean, Integer> right = dfs(node.right);

        boolean balanced = (left.getKey() && right.getKey() &&
                Math.abs(left.getValue() - right.getValue()) <= 1);

        return new Pair<Boolean, Integer>(balanced, 1 + Math.max(left.getValue(), right.getValue()));
    }*/

}

File: src/main/java/org/example/tree/BalancedBinaryTree.py
from typing import Optional

from org.example.tree.TreeNode import TreeNode


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root):  # returns a pair: [T/F, height of subtree]
            if not root: return [True, 0]

            left, right = dfs(root.left), dfs(root.right)
            balanced = (left[0] and right[0] and
                        abs(left[1] - right[1]) <= 1)
            return [balanced, 1 + max(left[1], right[1])]

        return dfs(root)[0]